napari/layers/shapes/_shape_list.py

```diff
@@ -353,6 +353,178 @@ def add(
             # Set z_order
             self._update_z_order()
 
+    def add_multiple(
```

Comment:
This method is getting pretty long. It might be nice to break it up into smaller util/helper functions and then `add_multiple()` can dispatch them. It would likely make this easier to read and also easier to share functions between `add()` and `add_multiple()`.

I understand this is a lot of carry over from the current implementation of the `add()` method. However, this seems like a good opportunity for a refactor.

Reply:
Along similar lines, do you think it makes sense to either get rid of `add()` or make `add()` a passthrough to `add_multiple()` (with any required coercion, etc.)?

Reply:
(Reactivating this PR, sorry for the delay)

Yes, ideally there should be only one method adding shapes. Currently, `add()` allows for adding shapes at a specific index, whereas `add_multiple()` will add consecutive indices at the end. It might be easy to merge those two, but I had (and still don't have) know real idea of what the purpose of the `shape_index` parameter has been and what side effects to expect, which is why I left `add()` as original.

Reply:
I think at the time `shape_index` seemed like a reasonable thing we might want - but if we ended up not using it then we could refactor it out.